src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,293 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
+    return ret;
+}
+
+static krb5_error_code
+certauth_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                   krb5_data cert, krb5_principal princ, void *opts,
+                   void *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_san, valid_eku;
+    struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    /* Decode cert into reqctx. */
+    ret = crypto_decode_der_cert(context, cert, req_opts->reqctx->cryptoctx);
+    if (ret)
+        goto err;
+
+    /* Verify the client SAN. */
+    ret = verify_client_san(context, req_opts->plgctx, req_opts->reqctx,
+                            req_opts->cb, req_opts->rock, princ, &valid_san);
+    if (ret)
+        goto err;
+
+    if (!valid_san) {
+        pkiDebug("%s: did not find an acceptable SAN in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+        goto err;
+    }
+
+    /* Verify the client EKU. */
+    ret = verify_client_eku(context, req_opts->plgctx, req_opts->reqctx,
+                            &valid_eku);
+    if (ret)
+        goto err;
+
+    if (!valid_eku) {
+        pkiDebug("%s: did not find an acceptable EKU in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_INCONSISTENT_KEY_PURPOSE;
+    }
+
+err:
+    return ret;
+}
+
+krb5_error_code
+certauth_pkinitsrv_initvt(krb5_context context, int maj_ver, int min_ver,
+                          krb5_plugin_vtable vtable)
+{
+    krb5_certauth_vtable vt;
+
+    if (maj_ver != 1)
+        return KRB5_PLUGIN_VER_NOTSUPP;
+    vt = (krb5_certauth_vtable)vtable;
+    vt->name = "pkinitsrv";
+    vt->authorize = certauth_authorize;
+    vt->init = NULL;
+    vt->fini = NULL;
+    vt->free_ind = NULL;
+    return 0;
+}
+
+/*
+ * Do certificate auth based on match expressions in the pkinit_cert_match
+ * attribute string.  An expression should be in the same form as those used for
+ * the pkinit_cert_match configuration option.  Multiple expressions are
+ * separated by ':'.  The certificate is authorized if all expressions are
+ * satisfied.
+ */
+static krb5_error_code
+dbmatch_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                  krb5_data cert, krb5_principal princ, void *opts,
+                  void *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    struct certauth_req_opts *req_opts = opts;
+    char *match = NULL, *matchcpy = NULL, *mp, *save;
+    krb5_boolean matched;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    ret = req_opts->cb->get_string(context, req_opts->rock,
+                                   "pkinit_cert_match", &match);
+    if (ret)
+        goto out;
+
+    /* Pass if there's no match expression. */
+    if (match == NULL) {
+        ret = KRB5_PLUGIN_NO_HANDLE;
+        goto out;
+    }
+    matchcpy = k5memdup0(match, strlen(match), &ret);
```

Comment:
Possibly moot, but this is more clearly written using strdup(), even though that doesn't set ret.  (I've considered adding a k5strdup helper for that reason, but haven't done so yet.)